homeassistant/components/cloudflare/config_flow.py

```diff
@@ -0,0 +1,180 @@
+import logging
+from typing import Dict, List, Optional
+
+from pycfdns import CloudflareUpdater
+from pycfdns.exceptions import (
+    CloudflareAuthenticationException,
+    CloudflareConnectionException,
+    CloudflareZoneException,
+)
+import voluptuous as vol
+
+from homeassistant.components import persistent_notification
+from homeassistant.config_entries import CONN_CLASS_CLOUD_PUSH, ConfigFlow
+from homeassistant.const import CONF_API_TOKEN, CONF_ZONE
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_RECORDS
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_TOKEN): str,
+    }
+)
+
+
+def _zone_schema(zones: Optional[List] = None):
+    """Zone selection schema."""
+    zones_list = []
+
+    if zones is not None:
+        zones_list = zones
+
+    return vol.Schema({vol.Required(CONF_ZONE): vol.In(zones_list)})
+
+
+def _records_schema(records: Optional[List] = None):
+    """Zone records selection schema."""
+    records_dict = {}
+
+    if records:
+        records_dict = {name: name for name in records}
+
+    return vol.Schema({vol.Required(CONF_RECORDS): cv.multi_select(records_dict)})
+
+
+async def validate_input(hass: HomeAssistant, data: Dict):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    zone = data.get(CONF_ZONE)
+    records = None
+
+    cfupdate = CloudflareUpdater(
+        async_get_clientsession(hass),
+        data[CONF_API_TOKEN],
+        zone,
+        [],
+    )
+
+    try:
+        zones = await cfupdate.get_zones()
+        if zone:
+            zone_id = await cfupdate.get_zone_id()
+            records = await cfupdate.get_zone_records(zone_id, "A")
+    except CloudflareConnectionException as error:
+        raise CannotConnect from error
+    except CloudflareAuthenticationException as error:
+        raise InvalidAuth from error
+    except CloudflareZoneException as error:
+        raise InvalidZone from error
+
+    return {"zones": zones, "records": records}
+
+
+class CloudflareConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Cloudflare."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_CLOUD_PUSH
+
+    def __init__(self):
+        """Initialize the Cloudflare config flow."""
+        self.cloudflare_config = {}
+        self.zones = None
+        self.records = None
+
+    async def async_step_user(self, user_input: Optional[Dict] = None):
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        assert self.hass
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)